mypyc/irbuild/specialize.py

```diff
@@ -135,13 +140,48 @@ def translate_safe_generator_call(
                     + [builder.accept(arg) for arg in expr.args[1:]]),
                 builder.node_type(expr), expr.line, expr.arg_kinds, expr.arg_names)
         else:
+            if callee.fullname == "builtins.tuple":
+                val = tuple_from_generator_helper(builder, expr.args[0])
+                if val is not None:
+                    return val
             return builder.call_refexpr_with_args(
                 expr, callee,
                 ([translate_list_comprehension(builder, expr.args[0])]
                     + [builder.accept(arg) for arg in expr.args[1:]]))
     return None
 
 
+def tuple_from_generator_helper(builder: IRBuilder,
+                                gen: GeneratorExpr) -> Optional[Value]:
+
+    if len(gen.sequences) == 1 and len(gen.condlists[0]) == 0:
+        # Currently we only optimize for simplest generator expression
+        if not is_list_rprimitive(builder.node_type(gen.sequences[0])):
```

Comment:
Here I limit the Generator with only one for-in loop and no conditions. Is it fine?

Reply:
Yeah, that's fine.